core/src/main/java/com/sequenceiq/cloudbreak/service/image/userdata/UserDataService.java

```diff
@@ -89,24 +101,54 @@ public void updateProxyConfig(Long stackId) {
         String gatewayUserData = userDataByInstanceGroup.get(InstanceGroupType.GATEWAY);
         String result = proxyConfigUserDataReplacer.replaceProxyConfigInUserDataByEnvCrn(gatewayUserData, stack.getEnvironmentCrn());
         userDataByInstanceGroup.put(InstanceGroupType.GATEWAY, result);
-        updateUserData(stack, userDataByInstanceGroup);
+        updateUserData(stack.getId(), userDataByInstanceGroup);
     }
 
-    private Map<InstanceGroupType, String> getUserData(Long stackId) {
+    public Map<InstanceGroupType, String> getUserData(Long stackId) {
         try {
             Image image = imageService.getImage(stackId);
-            return new HashMap<>(image.getUserdata());
+            if (image.getUserdata() == null || image.getUserdata().isEmpty()) {
+                Map<InstanceGroupType, String> map = new HashMap<>();
+                Optional<Userdata> userdataOptional = userdataRepository.findByStackId(stackId);
+                if (userdataOptional.isPresent()) {
+                    map.put(InstanceGroupType.CORE, userdataOptional.get().getCoreUserdata());
+                    map.put(InstanceGroupType.GATEWAY, userdataOptional.get().getGatewayUserdata());
+                }
+                return map;
+            } else {
+                return new HashMap<>(image.getUserdata());
+            }
         } catch (CloudbreakImageNotFoundException e) {
             throw convertToServiceException(e);
         }
     }
 
-    private void updateUserData(Stack stack, Map<InstanceGroupType, String> userdata) {
-        try {
-            imageService.decorateImageWithUserDataForStack(stack, userdata);
-        } catch (CloudbreakImageNotFoundException e) {
-            throw convertToServiceException(e);
+    public Userdata updateUserData(Long stackId, Map<InstanceGroupType, String> userdata) {
+        Stack stack = stackService.get(stackId);
+        Userdata result;
+
+        Optional<Userdata> userdataOptional = userdataRepository.findByStackId(stackId);
+
+        if (userdataOptional.isPresent()) {
+            Userdata existingUserData = userdataOptional.get();
+            String coreUserdataSecret = existingUserData.getCoreUserdata();
+            String gatewayUserdataSecret = existingUserData.getGatewayUserdata();
+
+            existingUserData.setCoreUserdata(userdata.get(InstanceGroupType.CORE));
+            existingUserData.setGatewayUserdata(userdata.get(InstanceGroupType.GATEWAY));
+            result = userdataRepository.save(existingUserData);
+
+            secretService.delete(coreUserdataSecret);
+            secretService.delete(gatewayUserdataSecret);
```

Comment:
The secret service delete method accepts a raw Secret JSON but you are supplying the content of the secret here.
Please make sure that the deletion indeed happens because statically analysing this part suggests otherwise.